internal/service/advancedcluster/data_source_advanced_cluster.go

```diff
@@ -272,95 +271,43 @@ func dataSourceRead(ctx context.Context, d *schema.ResourceData, meta any) diag.
 	clusterName := d.Get("name").(string)
 	useReplicationSpecPerShard := false
 	var replicationSpecs []map[string]any
-	var clusterID string
 
 	if v, ok := d.GetOk("use_replication_spec_per_shard"); ok {
 		useReplicationSpecPerShard = v.(bool)
 	}
 
-	if !useReplicationSpecPerShard {
-		clusterDescOld, resp, err := connV220240530.ClustersApi.GetCluster(ctx, projectID, clusterName).Execute()
-		if err != nil {
-			if resp != nil {
-				if resp.StatusCode == http.StatusNotFound {
-					return nil
-				}
-				if admin20240530.IsErrorCode(err, "ASYMMETRIC_SHARD_UNSUPPORTED") {
-					return diag.FromErr(fmt.Errorf("please add `use_replication_spec_per_shard = true` to your data source configuration to enable asymmetric shard support. Refer to documentation for more details. %s", err))
-				}
-			}
-			return diag.FromErr(fmt.Errorf(errorRead, clusterName, err))
-		}
-
-		clusterID = clusterDescOld.GetId()
-
-		if err := d.Set("disk_size_gb", clusterDescOld.GetDiskSizeGB()); err != nil {
-			return diag.FromErr(fmt.Errorf(ErrorClusterAdvancedSetting, "disk_size_gb", clusterName, err))
-		}
-		clusterDescNew, _, err := connV2.ClustersApi.GetCluster(ctx, projectID, clusterName).Execute()
-		if err != nil {
-			return diag.FromErr(fmt.Errorf(errorRead, clusterName, err))
-		}
-		if err := d.Set("replica_set_scaling_strategy", clusterDescNew.GetReplicaSetScalingStrategy()); err != nil {
-			return diag.FromErr(fmt.Errorf(ErrorClusterAdvancedSetting, "replica_set_scaling_strategy", clusterName, err))
-		}
-		if err := d.Set("redact_client_log_data", clusterDescNew.GetRedactClientLogData()); err != nil {
-			return diag.FromErr(fmt.Errorf(ErrorClusterAdvancedSetting, "redact_client_log_data", clusterName, err))
+	clusterDesc, resp, err := connV2.ClustersApi.GetCluster(ctx, projectID, clusterName).Execute()
+	if err != nil {
+		if resp != nil && resp.StatusCode == http.StatusNotFound {
+			return nil
 		}
-
-		zoneNameToZoneIDs, err := getZoneIDsFromNewAPI(clusterDescNew)
-		if err != nil {
+		return diag.FromErr(fmt.Errorf(errorRead, clusterName, err))
+	}
+	zoneNameToOldReplicationSpecMeta, err := GetReplicationSpecAttributesFromOldAPI(ctx, projectID, clusterName, connV220240530.ClustersApi)
+	if err != nil {
+		if apiError, ok := admin20240530.AsError(err); !ok {
 			return diag.FromErr(err)
+		} else if apiError.GetErrorCode() != "ASYMMETRIC_SHARD_UNSUPPORTED" || (apiError.GetErrorCode() == "ASYMMETRIC_SHARD_UNSUPPORTED" && useReplicationSpecPerShard) {
+			return diag.FromErr(fmt.Errorf(errorRead, clusterName, err))
+		} else if !useReplicationSpecPerShard {
+			return diag.FromErr(fmt.Errorf("please add `use_replication_spec_per_shard = true` to your data source configuration to enable asymmetric shard support. Refer to documentation for more details. %s", err))
 		}
+	}
```

Comment:
From this logic we are including a false positive for the case cluster is asymmetric and user defines useReplicationSpecPerShard

Reply:
Would follow same conditional as in plural data source:
```
if apiError, ok := admin20240530.AsError(err); !ok {
   return nil, diag.FromErr(err)
} else if apiError.GetErrorCode() == "ASYMMETRIC_SHARD_UNSUPPORTED" && !useReplicationSpecPerShard {
   continue
} else if apiError.GetErrorCode() != "ASYMMETRIC_SHARD_UNSUPPORTED" {
   return nil, diag.FromErr(err)
}
```
but for the middle case in the singlar data source we would throw an error with the more helpful messaging `please add use_replication_spec_per_shard = true to your ...`

Reply:
Changed, I think now it's clearer, let me know what you think